File: acrarium/src/main/java/com/faendir/acra/ui/component/dialog/AcrariumDialog.java
package com.faendir.acra.ui.component.dialog;

import com.faendir.acra.ui.component.Translatable;
import com.vaadin.flow.component.ClickEvent;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.button.ButtonVariant;
import com.vaadin.flow.component.dialog.Dialog;
import com.vaadin.flow.dom.Element;

import java.util.Optional;

public class AcrariumDialog extends Dialog {
    private final DialogContent content;

    public AcrariumDialog() {
        this.content = new DialogContent();
        super.add(content);
    }

    public void setHeader(String captionId, Object... params) {
        content.setHeader(Translatable.createH3(captionId, params));
    }

    public void setPositive(ComponentEventListener<ClickEvent<Button>> clickListener, String captionId, Object... params) {
        Translatable<Button> button = Translatable.createButton(event -> {
            close();
            clickListener.onComponentEvent(event);
        }, captionId, params);
        content.setPositive(button);
    }

    public Optional<Translatable<Button>> getPositive() {
        //noinspection unchecked
        return content.getElement().getChildren().filter(e -> "positive".equals(e.getAttribute("slot"))).findAny().flatMap(Element::getComponent).map(c -> (Translatable<Button>) c);
    }

    public void setNegative(String captionId, Object... params) {
        setNegative(e -> {
        }, captionId, params);
    }

    public void setNegative(ComponentEventListener<ClickEvent<Button>> clickListener, String captionId, Object... params) {
        Translatable<Button> button = Translatable.createButton(event -> {
            close();
            clickListener.onComponentEvent(event);
        }, captionId, params).with(b -> {
            b.removeThemeVariants(ButtonVariant.LUMO_PRIMARY);
            b.addThemeVariants(ButtonVariant.LUMO_TERTIARY);
        });
        content.setNegative(button);
    }

    @Override
    public void add(Component... components) {
        content.add(components);
    }

    @Override
    public void add(String text) {
        content.add(text);
    }

    @Override
    public void remove(Component... components) {
        content.remove(components);
    }

    @Override
    public void removeAll() {
        content.removeAll();
    }

    @Override
    public void addComponentAtIndex(int index, Component component) {
        content.addComponentAtIndex(index, component);
    }

    @Override
    public void addComponentAsFirst(Component component) {
        content.addComponentAsFirst(component);
    }
}

File: acrarium/src/main/kotlin/com/faendir/acra/ui/component/dialog/AcrariumDialog.kt
/*
 * (C) Copyright 2019-2021 Lukas Morawietz (https://github.com/F43nd1r)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.faendir.acra.ui.component.dialog

import com.faendir.acra.i18n.Messages
import com.faendir.acra.ui.component.Translatable
import com.vaadin.flow.component.ClickEvent
import com.vaadin.flow.component.Component
import com.vaadin.flow.component.Composite
import com.vaadin.flow.component.HasComponents
import com.vaadin.flow.component.button.Button
import com.vaadin.flow.component.button.ButtonVariant
import com.vaadin.flow.component.dialog.Dialog

open class AcrariumDialog : Composite<Dialog>(), HasComponents {
    private val dialogContent: DialogContent = DialogContent()

    init {
        content.add(dialogContent)
    }

    var isOpened: Boolean
        get() = content.isOpened
        set(value) {
            content.isOpened = value
        }

    fun open() {
        content.open()
    }

    fun close() {
        content.close()
    }

    fun header(captionId: String, vararg params: Any) {
        dialogContent.add(DialogContent.Slot.HEADER, Translatable.createH3(captionId, *params))
    }

    fun positiveAction(captionId: String, vararg params: Any, clickListener: (ClickEvent<Button>) -> Unit = {}) {
        dialogContent.add(DialogContent.Slot.POSITIVE, Translatable.createButton(captionId, *params) {
            close()
            clickListener(it)
        })
    }

    val positive: Translatable<Button>?
        get() = dialogContent.get(DialogContent.Slot.POSITIVE)
            .filterIsInstance<Translatable<Button>>()
            .firstOrNull()

    fun negativeAction(captionId: String, vararg params: Any, clickListener: (ClickEvent<Button>) -> Unit = {}) {
        dialogContent.add(DialogContent.Slot.NEGATIVE, Translatable.createButton(captionId, *params, theme = ButtonVariant.LUMO_TERTIARY) {
            close()
            clickListener(it)
        })
    }

    override fun add(vararg components: Component?) {
        dialogContent.add(*components)
    }

    override fun add(text: String?) {
        dialogContent.add(text)
    }

    override fun remove(vararg components: Component?) {
        dialogContent.remove(*components)
    }

    override fun removeAll() {
        dialogContent.removeAll()
    }

    override fun addComponentAtIndex(index: Int, component: Component?) {
        dialogContent.addComponentAtIndex(index, component)
    }

    override fun addComponentAsFirst(component: Component?) {
        dialogContent.addComponentAsFirst(component)
    }
}

fun AcrariumDialog.createButton(onCreateAction: AcrariumDialog.() -> Unit) {
    positiveAction(Messages.CREATE) { onCreateAction() }
    negativeAction(Messages.CANCEL)
}

fun AcrariumDialog.closeButton() {
    positiveAction(Messages.CLOSE)
}

fun AcrariumDialog.confirmButtons(onYesAction: AcrariumDialog.() -> Unit) {
    positiveAction(Messages.CONFIRM) { onYesAction() }
    negativeAction(Messages.CANCEL)
}